File: training_data/reflection-tuning/day24_part2_2016.py
from collections import deque
import itertools

def parse_input(filename):
    with open(filename, 'r') as f:
        grid = [list(line.strip()) for line in f]
    points = {}
    for i, row in enumerate(grid):
        for j, cell in enumerate(row):
            if cell.isdigit():
                points[int(cell)] = (i, j)
    return grid, points

def bfs(grid, start, end):
    queue = deque([(start, 0)])
    visited = set([start])
    while queue:
        (x, y), dist = queue.popleft()
        if (x, y) == end:
            return dist
        for dx, dy in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
            nx, ny = x + dx, y + dy
            if 0 <= nx < len(grid) and 0 <= ny < len(grid[0]) and grid[nx][ny] != '#' and (nx, ny) not in visited:
                queue.append(((nx, ny), dist + 1))
                visited.add((nx, ny))
    return float('inf')

def solve(grid, points):
    distances = {(i, j): bfs(grid, points[i], points[j]) for i in points for j in points if i != j}
    
    non_zero_points = [p for p in points if p != 0]
    min_distance = float('inf')
    min_distance_with_return = float('inf')

    for perm in itertools.permutations(non_zero_points):
        route = [0] + list(perm)
        distance = sum(distances[(route[i], route[i+1])] for i in range(len(route)-1))
        min_distance = min(min_distance, distance)
        min_distance_with_return = min(min_distance_with_return, distance + distances[(route[-1], 0)])

    return min_distance, min_distance_with_return

grid, points = parse_input("input.txt")
part1, part2 = solve(grid, points)
print(f"Part 1: {part1}")
print(f"Part 2: {part2}")

File: clojure/day24_part2_2016.clj

(ns air-duct-spelunking
  (:require [clojure.string :as str]
            [clojure.set :as set]))

(defn parse-map [input]
  (let [lines (str/split-lines input)
        height (count lines)
        width (count (first lines))
        grid (vec (map vec lines))
        locations (reduce (fn [acc row-idx]
                            (reduce (fn [acc col-idx]
                                      (let [c (get-in grid [row-idx col-idx])]
                                        (if (Character/isDigit c)
                                          (assoc acc (Integer/parseInt (str c)) [row-idx col-idx])
                                          acc)))
                                    acc
                                    (range width)))
                          {}
                          (range height))]
    {:grid grid
     :locations locations
     :height height
     :width width}))

(defn neighbors [[row col] {:keys [grid height width]}]
  (let [candidates [[(dec row) col] [(inc row) col] [row (dec col)] [row (inc col)]]
        valid-coords (fn [[r c]] (and (>= r 0) (< r height) (>= c 0) (< c width)))]
    (filter (fn [[r c]] (not= \# (get-in grid [r c])))
            (filter valid-coords candidates))))

(defn bfs [start end {:keys [grid] :as map-data}]
  (loop [queue (conj clojure.lang.PersistentQueue/EMPTY [start 0])
         visited #{start}]
    (if (empty? queue)
      nil ; No path found
      (let [[curr dist] (peek queue)
            queue (pop queue)]
        (if (= curr end)
          dist
          (let [new-neighbors (remove visited (neighbors curr map-data))]
            (recur (reduce conj queue (map (fn [n] [n (inc dist)]) new-neighbors))
                   (into visited new-neighbors))))))))

(defn calculate-distances [map-data]
  (let [locations (:locations map-data)
        location-ids (keys locations)]
    (reduce (fn [acc a]
              (reduce (fn [acc b]
                        (if (>= b a)
                          acc
                          (let [dist (bfs (get locations a) (get locations b) map-data)]
                            (assoc acc [a b] dist))))
                      acc
                      location-ids))
            {}
            location-ids)))

(defn shortest-path [distances locations return-to-start?]
  (let [non-zero-locations (disj (set (keys locations)) 0)
        all-permutations (->> non-zero-locations
                              (set/permutations)
                              (map (fn [perm] (cons 0 perm))))
        path-lengths (map (fn [path]
                             (reduce (fn [sum [a b]]
                                       (let [key (sort [a b])]
                                         (+ sum (or (distances key) (distances (reverse key))))))
                                     0
                                     (partition 2 1 path)))
                           all-permutations)]
    (if return-to-start?
      (reduce min (map (fn [len path]
                         (+ len (or (distances (sort [(last path) 0])) (distances (sort [0 (last path)])))))
                       path-lengths all-permutations))
      (apply min path-lengths))))

(defn solve [input return-to-start?]
  (let [map-data (parse-map input)
        distances (calculate-distances map-data)]
    (shortest-path distances (:locations map-data) return-to-start?)))

(defn -main [& args]
  (let [input (slurp "input.txt")]
    (println "Part 1:" (solve input false))
    (println "Part 2:" (solve input true))))
